File: static/local-js/urlValidation.js
/* global */

const URLValidation = (() => {
  const urlKeyPattern = /(^|[_-])url([_-]|$)/i
  const placeholderValues = new Set(['http://', 'https://'])

  function isUrlField (input) {
    if (!input) return false
    if (input.dataset && input.dataset.urlSkip === 'true') return false
    if (input.type && input.type.toLowerCase() === 'url') return true
    const key = (input.name || input.id || '').toLowerCase()
    return urlKeyPattern.test(key)
  }

  function isPlaceholder (value) {
    return placeholderValues.has(String(value || '').trim().toLowerCase())
  }

  function hasEmptyPort (text) {
    const match = String(text).match(/^[a-z][a-z0-9+.-]*:\/\/([^/?#]*)/i)
    if (!match) return false
    const authority = match[1]
    return authority.endsWith(':')
  }

  function validateValue (value) {
    if (value == null) return { valid: true }
    const text = String(value).trim()
    if (!text) return { valid: true }
    if (isPlaceholder(text)) return { valid: false, message: 'URL is incomplete.' }
    if (hasEmptyPort(text)) return { valid: false, message: 'URL port is missing after ":"' }
    let parsed
    try {
      parsed = new URL(text)
    } catch (err) {
      return { valid: false, message: 'Please enter a valid URL.' }
    }
    if (!parsed.protocol || !['http:', 'https:'].includes(parsed.protocol)) {
      return { valid: false, message: 'URL must start with http:// or https://.' }
    }
    if (!parsed.hostname) {
      return { valid: false, message: 'URL is missing a hostname.' }
    }
    if (parsed.port) {
      const portNum = Number(parsed.port)
      if (!Number.isInteger(portNum) || portNum < 1 || portNum > 65535) {
        return { valid: false, message: 'URL port must be between 1 and 65535.' }
      }
    }
    if (!isValidHostname(parsed.hostname)) {
      return { valid: false, message: 'URL hostname is invalid.' }
    }
    return { valid: true }
  }

  function ensureFeedback (input) {
    const parent = input.closest('.input-group') || input.parentElement
    if (!parent) return null
    let feedback = parent.querySelector('.invalid-feedback')
    if (!feedback) {
      feedback = document.createElement('div')
      feedback.className = 'invalid-feedback'
      parent.appendChild(feedback)
    }
    return feedback
  }

  function applyResult (input, result) {
    if (!input) return
    if (result.valid) {
      input.classList.remove('is-invalid')
      input.dataset.urlValid = 'true'
      const feedback = ensureFeedback(input)
      if (feedback) feedback.textContent = ''
      return
    }
    input.classList.add('is-invalid')
    input.dataset.urlValid = 'false'
    const feedback = ensureFeedback(input)
    if (feedback) feedback.textContent = result.message || 'Invalid URL.'
  }

  function bindInput (input) {
    if (!input || input.dataset.urlValidationBound === 'true') return
    input.dataset.urlValidationBound = 'true'
    const validate = () => {
      const result = validateValue(input.value)
      applyResult(input, result)
      return result.valid
    }
    input.addEventListener('input', validate)
    input.addEventListener('blur', validate)
    validate()
  }

  function attach (container = document) {
    const inputs = Array.from(container.querySelectorAll('input[type="text"], input[type="url"], input[type="search"], textarea'))
    inputs.forEach(input => {
      if (isUrlField(input)) {
        bindInput(input)
      }
    })
  }

  function validateAll (container = document) {
    let allValid = true
    const inputs = Array.from(container.querySelectorAll('[data-url-validation-bound="true"]'))
    inputs.forEach(input => {
      const result = validateValue(input.value)
      applyResult(input, result)
      if (!result.valid) allValid = false
    })
    return allValid
  }

  return {
    attach,
    validateAll
  }
})()

window.URLValidation = URLValidation

function isValidHostname (hostname) {
  const host = String(hostname || '').toLowerCase()
  if (!host) return false
  if (host === 'localhost') return true
  if (host.startsWith('.') || host.endsWith('.')) return false
  if (host.includes('..')) return false
  if (isIPv4(host) || isIPv6(host)) return true
  const labels = host.split('.')
  for (const label of labels) {
    if (!label.length || label.length > 63) return false
    if (label.startsWith('-') || label.endsWith('-')) return false
    if (!/^[a-z0-9-]+$/.test(label)) return false
  }
  if (labels.length > 1) {
    const tld = labels[labels.length - 1]
    if (tld.length < 2 || tld.length > 63) return false
    if (!/[a-z]/.test(tld)) return false
  }
  return true
}

function isIPv4 (host) {
  const parts = host.split('.')
  if (parts.length !== 4) return false
  return parts.every(part => {
    if (!/^\d+$/.test(part)) return false
    const num = Number(part)
    return num >= 0 && num <= 255
  })
}

function isIPv6 (host) {
  return /^[0-9a-f:]+$/i.test(host) && host.includes(':')
}

File: modules/url_validation.py
import ipaddress
import re
from urllib.parse import urlparse

URL_KEY_RE = re.compile(r"(^|[_-])url([_-]|$)")
BOOLEAN_TEXT_VALUES = {"true", "false", "on", "off", "yes", "no", "0", "1"}


def is_url_key(key):
    if not key:
        return False
    return bool(URL_KEY_RE.search(str(key)))


def is_placeholder(value):
    if value is None:
        return False
    text = str(value).strip().lower()
    return text in {"http://", "https://"}


def is_boolean_value(value):
    if isinstance(value, bool):
        return True
    if value is None:
        return False
    text = str(value).strip().lower()
    return text in BOOLEAN_TEXT_VALUES


def _is_local_address(hostname):
    host = str(hostname or "").strip().lower()
    if not host:
        return False
    if host in {"localhost", "localhost.localdomain"}:
        return True
    try:
        ip = ipaddress.ip_address(host)
    except ValueError:
        return False
    return ip.is_private or ip.is_loopback or ip.is_link_local or ip.is_reserved or ip.is_multicast or ip.is_unspecified


def validate_url(value, allow_local=True):
    if value is None:
        return True, None
    text = str(value).strip()
    if text == "":
        return True, None
    if is_placeholder(text):
        return False, "URL is incomplete."
    parsed = urlparse(text)
    if not parsed.scheme or parsed.scheme not in {"http", "https"}:
        return False, "URL must start with http:// or https://."
    if not parsed.netloc:
        return False, "URL is missing a hostname."
    if parsed.netloc.endswith(":"):
        return False, "URL port is missing after ':'"
    try:
        if parsed.port is not None:
            if parsed.port < 1 or parsed.port > 65535:
                return False, "URL port must be between 1 and 65535."
    except ValueError:
        return False, "URL port must be between 1 and 65535."
    hostname = parsed.hostname
    if not hostname:
        return False, "URL is missing a hostname."
    if not is_valid_hostname(hostname):
        return False, "URL hostname is invalid."
    if not allow_local and _is_local_address(hostname):
        return False, "URL hostname must be a public address."
    return True, None


def is_valid_hostname(hostname):
    host = str(hostname).lower()
    if not host:
        return False
    if host == "localhost":
        return True
    if host.startswith(".") or host.endswith("."):
        return False
    if ".." in host:
        return False
    try:
        ipaddress.ip_address(host)
        return True
    except ValueError:
        pass
    labels = host.split(".")
    for label in labels:
        if not 1 <= len(label) <= 63:
            return False
        if label.startswith("-") or label.endswith("-"):
            return False
        if not re.match(r"^[a-z0-9-]+$", label):
            return False
    if len(labels) > 1:
        tld = labels[-1]
        if len(tld) < 2 or len(tld) > 63:
            return False
        if not re.search(r"[a-z]", tld):
            return False
    return True


def validate_payload(payload):
    errors = []
    if payload is None:
        return errors

    items = payload.items() if isinstance(payload, dict) else []
    if hasattr(payload, "getlist"):
        items = []
        for key in payload.keys():
            values = payload.getlist(key)
            if not values:
                items.append((key, ""))
            elif len(values) == 1:
                items.append((key, values[0]))
            else:
                for val in values:
                    items.append((key, val))

    for key, value in items:
        if not is_url_key(key):
            continue
        if is_boolean_value(value):
            continue
        valid, message = validate_url(value)
        if not valid:
            errors.append(f"{key}: {message}")
    return errors
